vj4/handler/contest.py

```diff
@@ -337,3 +337,57 @@ class ContestCreateHandler(base.Handler, ContestStatusMixin):
     for pid in pids:
       await problem.set_hidden(self.domain_id, pid, True)
     self.json_or_redirect(self.reverse_url('contest_detail', tid=tid))
+
+
+@app.route('/contest/{tid}/edit', 'contest_edit')
+class ContestEditHandler(base.Handler, ContestStatusMixin):
+  @base.require_priv(builtin.PRIV_USER_PROFILE)
+  @base.route_argument
+  @base.sanitize
+  async def get(self, *, tid: objectid.ObjectId):
+    tdoc = await contest.get(self.domain_id, tid)
+    if not self.own(tdoc, builtin.PERM_EDIT_CONTEST_SELF):
+      self.check_perm(builtin.PERM_EDIT_CONTEST)
+    dt = pytz.utc.localize(tdoc['begin_at']).astimezone(self.timezone)
+    self.render('contest_edit.html', tdoc=tdoc,
+                date_text=dt.strftime('%Y-%m-%d'),
+                time_text=dt.strftime('%H:%M'))
+
+  @base.require_priv(builtin.PRIV_USER_PROFILE)
+  @base.require_perm(builtin.PERM_EDIT_PROBLEM)
+  @base.route_argument
+  @base.post_argument
+  @base.require_csrf_token
+  @base.sanitize
+  async def post(self, *, tid: objectid.ObjectId, title: str, content: str, rule: int,
+                 begin_at_date: str,
+                 begin_at_time: str,
+                 duration: float,
+                 pids: str):
+    tdoc = await contest.get(self.domain_id, tid)
+    if not self.own(tdoc, builtin.PERM_EDIT_CONTEST_SELF):
+      self.check_perm(builtin.PERM_EDIT_CONTEST)
+    try:
+      begin_at = datetime.datetime.strptime(begin_at_date + ' ' + begin_at_time, '%Y-%m-%d %H:%M')
```

Comment:
Is it possible to reuse the logic here? i.e. can we introduce a new type hint for data times?

Reply:
swx template :) :(